src/transaction_sender/service.zig

```diff
@@ -0,0 +1,349 @@
+    receive_transactions_handle.join();
+    process_transactions_handle.join();
+}
+
+fn receiveTransactionsThread(
+    allocator: std.mem.Allocator,
+    receiver: *Channel(TransactionInfo),
+    sender: *Channel(std.ArrayList(Packet)),
+    config: Config,
+    cluster_info_rw: *RwMux(LeaderInfo),
+    transaction_pool_rw: *RwMux(TransactionPool),
+    exit: *AtomicBool,
+    logger: Logger,
+) !void {
+    errdefer exit.store(true, .unordered);
+
+    var last_batch_sent = try Instant.now();
+
+    var transaction_batch = TransactionPool.init(allocator);
+    defer transaction_batch.deinit();
+
+    while (!exit.load(.unordered)) {
+        const transaction = receiver.receive() orelse break;
+
+        if (!transaction_batch.contains(transaction.signature)) {
+            const transaction_pool: *const TransactionPool, var transaction_pool_lock = transaction_pool_rw.readWithLock();
+            defer transaction_pool_lock.unlock();
+
+            if (!transaction_pool.contains(transaction.signature)) {
+                try transaction_batch.put(transaction.signature, transaction);
+            }
+        }
+
+        if (transaction_batch.count() >= config.batch_size or
+            (transaction_batch.count() > 0 and
+            (try last_batch_sent.elapsed()).asNanos() >= config.batch_send_rate.asNanos()))
+        {
+            try sendTransactions(
+                allocator,
+                sender,
+                config,
+                cluster_info_rw,
+                transaction_batch.values(),
+            );
+            last_batch_sent = try Instant.now();
+
+            var transaction_pool: *TransactionPool, var transaction_pool_lock = transaction_pool_rw.writeWithLock();
+            defer transaction_pool_lock.unlock();
+
+            logger.infof("Adding {d} new transactions to pool.", .{transaction_batch.count()});
+            for (transaction_batch.values()) |_tx| {
+                if (transaction_pool.count() >= config.pool_max_size) {
+                    logger.warnf("Transaction pool is full, dropping transaction: signature={s}", .{_tx.signature});
+                    continue;
+                }
+                var tx = _tx; // Is there a nicer way to do this?
+                tx.last_sent_time = last_batch_sent;
+                try transaction_pool.put(tx.signature, tx);
+            }
+            transaction_batch.clearRetainingCapacity();
+        }
+    }
+}
+
+fn processTransactionsThread(
+    allocator: std.mem.Allocator,
+    sender: *Channel(std.ArrayList(Packet)),
+    config: Config,
+    cluster_info_rw: *RwMux(LeaderInfo),
+    transaction_pool_rw: *RwMux(TransactionPool),
+    exit: *AtomicBool,
+    logger: Logger,
+) !void {
+    errdefer exit.store(true, .unordered);
+
+    while (!exit.load(.unordered)) {
+        std.time.sleep(config.pool_process_rate.asNanos());
+
+        var transaction_pool: *TransactionPool, var transaction_pool_lock = transaction_pool_rw.writeWithLock();
```

Comment:
this write-lock is held for much longer than we need (only really needed during `removeSwap`) - probs want to figure out a way to hold locks for as small as time as possible

Reply:
3981760